### ISSUE TITLE
[Question] learning <handson-ml2> with 1st edition book

### ISSUE THREAD

**Issue by xujiboy** [ageron/handson-ml2#7]
Hi Aurelien,

First of all this is a question instead of an issue; please let me know if this is not the right place to ask a book-related question and redirect me to the right place :)

I bought the 1st edition of the book and so far have been enjoying it very much. As TensorFlow 2.0 (TF) came out, from the ipynbs on this repo it seems that the 2nd edition will focus on taking advantage of the newer version of TF. I noticed that Chapter 9 (Up and Running with TensorFlow) of the 1st edition book is gone in the 2nd edition, may I know if this is because the low-level compute-graph construction is no longer applicable in TF2.0? And for a reader/learner like me, before the release of 2nd edition in Aug 2019, do you have any suggestions / advices on trying to use the 1st edition book to learn with latest version of TF? Should I skip particular chapters? Your advices would be greatly appreciated.

Best Regards,
Ji

**Comment by ageron** [comment]
Hi @xujiboy ,
Thanks for your question! It's fine to ask questions here. If you notice errors in the code, github is the right place too. However, for errors in the text, then please file an erratum at https://homl.info/errata (1st edition) or https://homl.info/errata2 (2nd edition).

The chapter 9 from the 1st edition (Up and Running with TensorFlow) was updated to TF 2.0 in the second edition, and moved to chapter 12 (Custom models and training with TensorFlow).

Indeed, manually creating a graph, using sessions, variable scopes and collections, all of this is gone in TensorFlow 2.0. Also, tf.layers is gone, a large part of the API is cleaned up, there is no more tf.contrib, Keras is now central (most TF 2.0 code will be essentially Keras code). In short, it's really pretty different.

If you plan to continue to use TF 1.x for a while (for example because you have existing code, or your company has a lot of TF 1.x code), then you can just stick to TF 1.x for now, there's no rush, and the 1st edition of my book is still relevant for that. However, if you plan to migrate soon, or you just want to go straight to TF 2.0, then I would recommend getting the 2nd edition of my book. The early release is available at https://homl.info/er2 (you need to create a Safari account, and there's a free trial if you want).

Hope this helps!

**Comment by xujiboy** [comment]
Hi @ageron ,

Thank you so much for the detailed answer and advices. Now it's clear to me that the TF content in the newer edition of the book is what I need. As software frameworks evolves fast I intend to keep up with the latest and greatest. Before the hardcopy release I will give the Safari free trial a shot and access the early release of the 2nd edition there. 

Best
